### PR TITLE
feat: registry instance SDK

### PR DESCRIPTION
**Pull request by wtrocki**

### CODE REVIEW

**Comment by wtrocki** [comment]
@craicoverflow I have followed contributing guide to add new service but I might miss something. 
Looking for some hints what else to change (as this PR is still not ready for proper review).

My questions:

1. How to build SDK or examples? There is no make target etc.
2. Can I generate initial SDK locally somehow? I seen some docs before but not sure where it is right now :)

**Comment by craicoverflow** [comment]
> @craicoverflow I have followed contributing guide to add new service but I might miss something.
> Looking for some hints what else to change (as this PR is still not ready for proper review).
> 
> My questions:
> 
> 1. How to build SDK or examples? There is no make target etc.

The SDK does not need to be compiled, but you can import `go get` it by the branch like `go get github.com/owner/repo@feature-branch`

> 2. Can I generate initial SDK locally somehow? I seen some docs before but not sure where it is right now :)

You can generate locally like so: https://github.com/redhat-developer/app-services-sdk-go/tree/main/internal/apigen#generate-api-client

Hope this helps.

**Comment by wtrocki** [comment]
Oh.. I found it. Problem was that there was a readme in an internal folder that wasn't linked anywhere. 
Adding this to contributing guide

**Comment by wtrocki** [comment]
Got error:

> 2021/07/26 14:38:59 error unmarshaling JSON: failed to unmarshal property "paths" (*openapi3.Paths): failed to unmarshal extension properties: json: cannot unmarshal string into Go value of type map[string]json.RawMessage ("/apis/registry/v2")

Any idea what command is causing it - is that codegen or validator?

**Comment by wtrocki** [comment]
```
[wtrocki@graphapi app-services-sdk-go (registry-instance ✗)]$ api-generate --client-id="registryinstance/v1" --generator=go --templates-dir="./scripts/templates" --repo-metadata ".config/api-client-metadata.json"
zsh: command not found: api-generate
```

**Comment by craicoverflow** [comment]
> Got error:
> 
> > 2021/07/26 14:38:59 error unmarshaling JSON: failed to unmarshal property "paths" (*openapi3.Paths): failed to unmarshal extension properties: json: cannot unmarshal string into Go value of type map[string]json.RawMessage ("/apis/registry/v2")
> 
> Any idea what command is causing it - is that codegen or validator?

It looks to be caused by this: https://github.com/redhat-developer/app-services-sdk-go/blob/b8c5d4e22c23de66e6762a8776a385b519c67e4c/.openapi/registry-instance.json#L2186

Is that a valid top-level property of `paths`?

Also, not sure if you are aware that this OpenAPI schema has multiple objects defined:

https://github.com/redhat-developer/app-services-sdk-go/blob/b8c5d4e22c23de66e6762a8776a385b519c67e4c/.openapi/registry-instance.json#L3150

**Comment by wtrocki** [comment]
Initially I wanted to extend codegen to select only handful of tags we need, however after reviewing requirements I see that we still need to use some of the admin paths in the CLI meaning that this will not be sufficient to resolve our problems.

Problems I got with the OpenAPI: 

- API cannot have more than 1 tag for API spec. (specific for golang)
All endpoints defining multiple tags causing duplicates in the API - not only files are generated twice they are not compiled properly. 
The solution for me was to remove 2 tags and keep only one. That is something that we can discuss adding to master although it is simple to document and can stay like this in our SDK considering that open API is not changing that much.

-  x-codegen-contextRoot - seems to be causing problems for the generator and when I think about it 
Solution for me was to simply remove it.
However my take is that path is not the right place to put it. This extension should be defined in servers or objects root

**Comment by wtrocki** [comment]
Ready for review. @craicoverflow how do you run examples?

**Comment by craicoverflow** [comment]
> Ready for review. @craicoverflow how do you run examples?

The examples files might need a new environment variable to configure the API URL, but you can do this:

```shell
ACCESS_TOKEN="fdsfcdwff...sdfdsd" go run ./internal/examples/kafkamgmt/kafka_mgmt.go
```

It will not work, because the API URL is not set. Perhaps it should be hard-coded.

**Comment by wtrocki** [comment]
Cool adding examples and readme to examples

**Comment by wtrocki** [comment]
Good to review

**Comment by wtrocki** [comment]
@craicoverflow number of general SDK improvements in this PR. I would sleep well if you take look before merge.

File: registryinstance/apiv1internal/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+## Endpoints
+
+For a full list of the available endpoints, see [Documentation for API Endpoints](./client/README.md#documentation-for-api-endpoints).
+
+## Local development
+
+Registry API requires local patch in order to be valid. 
+Plese execute manually
```
Comment (craicoverflow) [inline registryinstance/apiv1internal/README.md:58]:
```suggestion
Please execute manually:
```

File: registryinstance/apiv1internal/api_client.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package registryinstance
+
+import (
+	"github.com/redhat-developer/app-services-sdk-go/internal"
+
+	apiv1 "github.com/redhat-developer/app-services-sdk-go/registryinstance/apiv1internal/client"
+)
+
+// APIConfig defines the available configuration options
```
Comment (craicoverflow) [inline registryinstance/apiv1internal/api_client.go:9]:
```suggestion
// Config defines the available configuration options
```

File: registryinstance/apiv1internal/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+For a full list of the available endpoints, see [Documentation for API Endpoints](./client/README.md#documentation-for-api-endpoints).
+
+## Local development
+
+Registry API requires local patch in order to be valid. 
+Plese execute manually 
+
+```bash
+./scripts/patch_apis.sh
```
Comment (craicoverflow) [inline registryinstance/apiv1internal/README.md:61]: > Registry API requires local patch in order to be valid.

Should this script also be executed within the GitHub Action to ensure the patch will be auto-applied?

**Review by craicoverflow — APPROVED** [review]
I have one comment, other than that all good.

File: registryinstance/apiv1internal/README.md (hunk 3 above)
Comment (wtrocki) [reply registryinstance/apiv1internal/README.md:61]: The registry integration is manual due to api being stagnant and the team providing their own SDKs for java.
There is no CI/CD involved here.

**Pull request merged**